display/download.go

```diff
@@ -83,56 +86,103 @@ func dlInit() {
 	})
 }
 
+func getMediaHandler(resp *gemini.Response) config.MediaHandler {
+	def := config.MediaHandler{
+		Cmd:      nil,
+		NoPrompt: false,
+	}
+
+	mediatype, _, err := mime.ParseMediaType(resp.Meta)
+	if err != nil {
+		return def
+	}
+
+	if ret, ok := config.MediaHandlers[mediatype]; ok {
+		return ret
+	}
+
+	splitType := strings.Split(mediatype, "/")
+	if ret, ok := config.MediaHandlers[splitType[0]]; ok {
+		return ret
+	}
+
+	if ret, ok := config.MediaHandlers["*"]; ok {
+		return ret
+	}
+
+	return def
+}
+
 // dlChoice displays the download choice modal and acts on the user's choice.
 // It should run in a goroutine.
 func dlChoice(text, u string, resp *gemini.Response) {
 	defer resp.Body.Close()
 
-	parsed, err := url.Parse(u)
-	if err != nil {
-		Error("URL Error", err.Error())
-		return
-	}
+	mediaHandler := getMediaHandler(resp)
+	var choice string
 
-	dlChoiceModal.SetText(text)
-	tabPages.ShowPage("dlChoice")
-	tabPages.SendToFront("dlChoice")
-	App.SetFocus(dlChoiceModal)
-	App.Draw()
+	if mediaHandler.NoPrompt {
+		choice = "Open"
+	} else {
+		dlChoiceModal.SetText(text)
+		tabPages.ShowPage("dlChoice")
+		tabPages.SendToFront("dlChoice")
+		App.SetFocus(dlChoiceModal)
+		App.Draw()
+		choice = <-dlChoiceCh
+	}
 
-	choice := <-dlChoiceCh
 	if choice == "Download" {
 		tabPages.HidePage("dlChoice")
 		App.Draw()
-		downloadURL(u, resp)
+		downloadURL(config.DownloadsDir, u, resp)
 		return
 	}
-	if choice == "Open in portal" {
-		// Open in mozz's proxy
-		portalURL := u
-		if parsed.RawQuery != "" {
-			// Remove query and add encoded version on the end
-			query := parsed.RawQuery
-			parsed.RawQuery = ""
-			portalURL = parsed.String() + "%3F" + query
-		}
-		portalURL = strings.TrimPrefix(portalURL, "gemini://") + "?raw=1"
-		ok := handleHTTP("https://portal.mozz.us/gemini/"+portalURL, false)
-		if ok {
-			tabPages.SwitchToPage(strconv.Itoa(curTab))
-			App.SetFocus(tabs[curTab].view)
-			App.Draw()
-		}
+	if choice == "Open" {
+		tabPages.HidePage("dlChoice")
+		App.Draw()
+		downloadAndOpen(u, resp)
 		return
 	}
 	tabPages.SwitchToPage(strconv.Itoa(curTab))
 	App.SetFocus(tabs[curTab].view)
 	App.Draw()
 }
 
+// downloadAndOpen performs the same actions as downloadURL except it also opens the file.
+// If there is no system viewer configured for the particular mediatype, it opens it
+// with the default system viewer.
+func downloadAndOpen(u string, resp *gemini.Response) {
+	mediaHandler := getMediaHandler(resp)
+	path := downloadURL(config.TempDownloadsDir, u, resp)
+	if path == "" {
+		return
+	}
+	if mediaHandler.Cmd == nil {
+		// Open with system default viewer
+		_, err := webbrowser.Open(path)
+		if err != nil {
+			Error("System Viewer Error", err.Error())
+			return
+		}
+		dlModal.SetText("Opened in system default viewer")
```

Comment:
```suggestion
		dlModal.SetText("Opened in default system viewer")
```
This sounds more natural to me.

Also I think that this message might be better suited to an `Info`. Do you think you could close the `dlModal` (See lines 80-85 for how to do that) and display this using `Info()` instead? And do the same on line 176.

If that's too messy or you think this method makes more sense let me know.

Reply:
Aha. I'm glad you have an example of how to do this. I was wondering how to close a modal without user interaction. 

When I was first doing this, I was wondering if the user needed any notification at all. Presumably they will be aware because the download message will have disappeared, and their file has opened up. What do you think?

Reply:
I think that displaying an Info message is good to let the user know Amfora has done its job. Maybe the application that is being called has failed silently, or is taking a long time to open. Displaying an Info shows that any issue that occurs isn't with Amfora.